### PR TITLE
Updated resource auditing to detect invalid mirrors when using --online

### PR DESCRIPTION
**Pull request by broder**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This should hopefully cover #935 by checking mirrors for 404s when `brew audit --online` is used. I haven't written tests as there didn't appear to be a sensible way of stubbing `curl` although let me know if you have any suggestions!

### CODE REVIEW

File: Library/Homebrew/dev-cmd/audit.rb [hunk 1]
```diff
@@ -1374,6 +1377,15 @@ def audit_urls
       next unless u =~ %r{https?://(?:central|repo\d+)\.maven\.org/maven2/(.+)$}
       problem "#{u} should be `https://search.maven.org/remotecontent?filepath=#{$1}`"
     end
+
+    return unless @online
+    urls.each do |u|
+      begin
+        nostdout { curl "--connect-timeout", "15", "-o", "/dev/null", "-r", "0-0", u }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: Prefer `--output` and `--range` here for readability. I wonder if this is preferable to e.g. a `HEAD` request? Any thoughts?

File: Library/Homebrew/dev-cmd/audit.rb [hunk 2]
```diff
@@ -618,11 +618,13 @@ def audit_specs
     %w[Stable Devel HEAD].each do |name|
       next unless spec = formula.send(name.downcase)
 
-      ra = ResourceAuditor.new(spec).audit
+      options = { online: @online }
+
+      ra = ResourceAuditor.new(spec, options).audit
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: `ra = ResourceAuditor.new(spec, online: @online).audit`

File: Library/Homebrew/dev-cmd/audit.rb [hunk 3]
```diff
@@ -1374,6 +1377,15 @@ def audit_urls
       next unless u =~ %r{https?://(?:central|repo\d+)\.maven\.org/maven2/(.+)$}
       problem "#{u} should be `https://search.maven.org/remotecontent?filepath=#{$1}`"
     end
+
+    return unless @online
+    urls.each do |u|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: `urls.each do |url|`

File: Library/Homebrew/dev-cmd/audit.rb (hunk 1 above)
Comment (broder) [inline Library/Homebrew/dev-cmd/audit.rb]: Good idea with the arguments!

From what I've read, a `HEAD` request would be preferable but some mirrors seem to block `HEAD` requests (one example being github releases e.g. https://github.com/admesh/admesh/releases/download/v0.98.2/admesh-0.98.2.tar.gz). I'm hoping that requesting the first byte will be more reliable.

**Comment by MikeMcQuaid** [comment]
This isn't working quite as expected:
```
mikebook # brew audit --online wget
wget:
  * HEAD: The mirror git://git.savannah.gnu.org/wget.git is not reachable (curl exit code 1)
Error: 1 problem in 1 formula
```

Additionally the failure case could be tweaked:
```
imagemagick:
  * Stable: The mirror https://www.imagemagick.org/download/ImageMagick-6.9.6-7222.tar.xz is not reachable (curl exit code 22)
```

Might be nice to note the HTTP status code rather than curl exit code, here.

It should probably know to not test non-http URLs. Otherwise this makes sense so far 👍

File: Library/Homebrew/dev-cmd/audit.rb [hunk 4]
```diff
@@ -1374,6 +1375,25 @@ def audit_urls
...
       problem "#{u} should be `https://search.maven.org/remotecontent?filepath=#{$1}`"
     end
+
+    return unless @online
+    urls.each do |url|
+      if url.start_with? "http", "ftp"
+        status_code, _, _ = curl_output "--connect-timeout", "15", "--output", "/dev/null", "--range", "0-0", \
+                                        "--write-out", "%{http_code}", url
+        unless status_code.start_with? "20"
+          problem "The mirror #{url} is not reachable (HTTP status code #{status_code})"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: Final (hopefully) nit: these are necessarily all mirrors so Just go for `problem "The URL #{url}` here and below.

File: Library/Homebrew/dev-cmd/audit.rb [hunk 5]
```diff
@@ -1374,6 +1375,25 @@ def audit_urls
       next unless u =~ %r{https?://(?:central|repo\d+)\.maven\.org/maven2/(.+)$}
       problem "#{u} should be `https://search.maven.org/remotecontent?filepath=#{$1}`"
     end
+
+    return unless @online
+    urls.each do |url|
+      if url.start_with? "http", "ftp"
+        status_code, _, _ = curl_output "--connect-timeout", "15", "--output", "/dev/null", "--range", "0-0", \
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: `brew style` tweak: https://travis-ci.org/Homebrew/brew/jobs/183090304

**Comment by broder** [comment]
Hey @MikeMcQuaid, thanks for your review / comments :) Is there anything still outstanding for this PR?

**Comment by MikeMcQuaid** [comment]
@broder Take a look at https://github.com/Homebrew/brew/pull/1725; once it's merged we may want to unify the approach between these (and perhaps the `curl` calls into a separate function in `audit`)?

**Comment by MikeMcQuaid** [comment]
@broder https://github.com/Homebrew/brew/pull/1725 been merged so you can use it here.

File: Library/Homebrew/dev-cmd/audit.rb [hunk 6]
```diff
@@ -1427,6 +1401,24 @@ def audit_urls
       next unless u =~ %r{https?://(?:central|repo\d+)\.maven\.org/maven2/(.+)$}
       problem "#{u} should be `https://search.maven.org/remotecontent?filepath=#{$1}`"
     end
+
+    return unless @online
+    urls.each do |url|
+      if url.start_with? "http", "ftp"
+        status_code = url_status_code url
+        unless status_code.start_with? "2"
```
Comment (broder) [inline Library/Homebrew/dev-cmd/audit.rb]: An example which returns a non `20x` code is `brew audit --online rc` which returns 226.

**Comment by broder** [comment]
Thanks @MikeMcQuaid! I've incorporated your changes now.

File: Library/Homebrew/utils/curl.rb [hunk 7]
```diff
@@ -40,3 +40,30 @@ def curl_output(*args)
     [stdout.read, stderr.read, wait_thread.value]
   end
 end
+
+def url_status_code(url, range: false, user_agent: :default)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/curl.rb]: Would rather leave this in `audit` until it's used elsewhere, I think. It also makes it easier to see what the changes have been (if any).

File: Library/Homebrew/dev-cmd/audit.rb (hunk 6 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: 🆒 👍

**Comment by MikeMcQuaid** [comment]
>   * HEAD: The URL http://git.imagemagick.org/repos/ImageMagick.git is not reachable (HTTP status code 406)

Wondering if we should set a user agent matching Git for HEAD URLs?

**Comment by broder** [comment]
Turns out that it's not based on User-Agent - git goes to some different URLs when you clone (https://git-scm.com/book/en/v2/Git-Internals-Transfer-Protocols). TIL!

I've moved the git check to the start and added `url.end_with?(".git")` which should hopefully fix issues similar to the ImageMagick one.

Thanks!

**Comment by alyssais** [comment]
@broder you could consider using the [existing logic](https://github.com/Homebrew/brew/blob/35045b2934d94eabe302693a05b12fb530827454/Library/Homebrew/download_strategy.rb#L1134-L1193) Homebrew has to detect if a URL should be fetched with Git or not.

**Comment by broder** [comment]
Nice, good idea!

**Comment by MikeMcQuaid** [comment]
Failing on a `file://` audit: https://travis-ci.org/Homebrew/brew/jobs/194999736

**Comment by MikeMcQuaid** [comment]
Want to do some more widespread testing and then 🚢 this. @broder If you detail what you've used as test cases that'd be helpful, thanks!

**Comment by MikeMcQuaid** [comment]
@broder This now needs rebased. Once you've done that this should be good to 🚢

**Comment by MikeMcQuaid** [comment]
> I don't think we should audit mirror URLs unless we're only auditing the server is still there

I'm not sure what this means. We should ignore mirror 404s?

**Comment by ilovezfs** [comment]
Absolutely, unless it's the server itself that is 404.

**Comment by MikeMcQuaid** [comment]
> Absolutely, unless it's the server itself that is 404.

@ilovezfs How do you differentiate between a server 404 and a mirror 404?

Seems like a bad idea to me to ignore mirror 404s but I'm willing to compromise to get this merged. Can we at least do them on `--strict`?

**Comment by ilovezfs** [comment]
On strict seems fine, but on non-strict I *would* like it to flag 404'ing servers!

**Comment by MikeMcQuaid** [comment]
@ilovezfs I don't think that's going to be straightforward. I don't think you can assume the root directory of a mirror will produce a HTTP success code.

**Comment by ilovezfs** [comment]
So far everything I've checked has been 200 except pkgs.fedoraproject.org, which gives 301.

**Comment by MikeMcQuaid** [comment]
@ilovezfs Now only checking mirrors with `--strict` and checking homepages (which shares logic so a bunch could be removed).

**Comment by ilovezfs** [comment]
Thanks.

**Comment by MikeMcQuaid** [comment]
Thanks again for your work here, @broder!

**Comment by ilovezfs** [comment]
@MikeMcQuaid as noted [here](https://github.com/Homebrew/homebrew-core/pull/12736#issuecomment-296256403) and [here](https://github.com/Homebrew/homebrew-core/pull/12739#issuecomment-296360113) this is actually still checking mirrors even when `--strict` is passed due to the other audit that requires GNU mirrors to be used for the primary URL.

**Comment by MikeMcQuaid** [comment]
Discussed with @ilovezfs in Slack and we're going to fiddle with GNU mirrors rather than change the audit: https://github.com/Homebrew/brew/pull/2526

**Pull request merged**